lib/main.rb

```diff
@@ -0,0 +1,82 @@
+module Enumerable
+  def my_each
+    element = 0
+    while element < length
+      yield(self[element])
+      element += 1
+    end
+    self
+  end
+
+  def my_each_with_index
+    element = 0
+    while element < length
+      yield(self[element], element)
+      element += 1
+    end
+  end
+
+  def my_select
+    new_array = []
+    my_each do |a|
+      yield(a) ? new_array.push(a) : new_array
+    end
+  end
+
+  def my_all?
+    my_each do |a|
+      return false unless yield(a)
+    end
+    true
+  end
+
+  def my_any?
+    exist = false
+    my_each do |element|
+      exist = yield(element)
+      break if exist
+    end
+    exist
+  end
+
+  def my_none?
+    (my_any? { |i| yield(i) == true }) != true
+  end
+
+  def my_count(number = nil)
+    total = 0
+    if number
+      my_each { |element| total += 1 if element == number }
+    elsif block_given?
+      my_each { |element| total += 1 if yield(element) }
+    else
+      total = length
+    end
+    total
+  end
+
+  def my_map(proc = nil)
+    new_arr = []
+    if proc
+      my_each do |element|
+        new_arr << proc.call(element)
+      end
+    elsif proc.nil? && block_given?
+      my_each do |element|
+        new_arr << yield(element)
+      end
+    end
+    new_arr
+  end
+
+  def my_inject(s_t)
+    my_each { |i| s_t = yield(s_t, i) }
+    s_t
+  end
+end
```

Comment:
- [ ] should be identical and returns the same thing as ruby's [`inject`](https://ruby-doc.org/core-2.6.4/Enumerable.html#method-i-inject). ([Screenshot](https://gitlab.com/microverse/guides/projects/requirements_screenshots/raw/master/images/ruby/advanced_building_blocks_enumerable/my_inject.png) from [Odin](https://www.theodinproject.com/courses/ruby-programming/lessons/advanced-building-blocks#assignment-2))
- [ ] my_inject combines all elements of enum by applying a binary operation, specified by a block
```ruby
  range = Range.new(5, 50)
  p range.my_inject(4) { |prod, n| prod * n } # returns this error
  # NoMethodError:
     #  undefined method `length' for 5..50:Range
    # ./enumerables.rb:76:in `my_inject'
```
- [ ] my_inject when a symbol is specified combines each element of the collection by applying the symbol as a named method
```ruby
  array = [1,7,8,89,9,6,5,12,4,2,5]
  p array.my_inject(:+) == array.inject(:+) # return this error
  # LocalJumpError:
     # no block given (yield)
    # ./enumerables.rb:78:in `block in my_inject'
    # ./enumerables.rb:76:in `each'
    # ./enumerables.rb:76:in `my_inject'
```